### PR TITLE
PHOENIX-6622 Transform monitor

### PR DESCRIPTION
**Pull request by gokceni**
TransformMonitor to monitor transform tool result and do orchestration

### CODE REVIEW

**Comment by gokceni** [comment]
@gjacoby126 

I blocked transform working on Transactional tables because we will need IndexRegionObserver to copy live transaction to the new table. FYI @stoty

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/transform/TransformTool.java [hunk 1]
```diff
@@ -775,30 +798,58 @@ public void resumeTransform(String[] args, CommandLine cmdLine) throws Exception
         }
 
         runTransform(args, cmdLine);
+
+        // Check if we already have a TransformMonitor task. If we do, remove those and start a new monitor
+        List<Task.TaskRecord> taskRecordList = Task.queryTaskTable(connection, null);
+        for (Task.TaskRecord taskRecord : taskRecordList) {
+            if (taskRecord.getTaskType() == PTable.TaskType.TRANSFORM_MONITOR) {
+                Task.deleteTask(connection.unwrap(PhoenixConnection.class), PTable.TaskType.TRANSFORM_MONITOR, taskRecord.getTimeStamp(), taskRecord.getTenantId(),
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/transform/TransformTool.java:806]: Isn't each TransformMonitor Task only meant for one particular transform record? Shouldn't we only delete the task if taskRecord matches the transform record we're interested in?

**Review by gjacoby126 — COMMENTED** [review]
Halfway through first pass

File: phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java [hunk 2]
```diff
@@ -2401,11 +2404,15 @@ private PTable createTableInternal(CreateTableStatement statement, byte[][] spli
                         linkStatement.setString(4, physicalName.getString());
                         linkStatement.setByte(5, LinkType.PHYSICAL_TABLE.getSerializedValue());
                         if (tableType == PTableType.VIEW) {
-                            PTable logicalTable = connection.getTable(new PTableKey(null, physicalName.getString()
-                                    .replace(QueryConstants.NAMESPACE_SEPARATOR, QueryConstants.NAME_SEPARATOR)));
+                            if (parent.getType() == PTableType.TABLE) {
+                                linkStatement.setString(4, SchemaUtil.getTableName(parent.getSchemaName().getString(),parent.getTableName().getString()));
+                                linkStatement.setLong(6, parent.getSequenceNumber());
+                            } else { //This is a grandchild view, find the physical base table
+                                PTable logicalTable = connection.getTable(new PTableKey(null, physicalName.getString().replace(QueryConstants.NAMESPACE_SEPARATOR, QueryConstants.NAME_SEPARATOR)));
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java:2411]: Think you could use SchemaUtil.gePhysicalName here, or something similar? If there's not exactly what you need already, probably cleaner to put a new SchemaUtil in.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/transform/Transform.java [hunk 3]
```diff
@@ -379,71 +421,265 @@ public static void upsertTransform(
...
+                connection.commit();
+                batchSize = 0;
+            }
+
+            connection.unwrap(PhoenixConnection.class).getQueryServices().clearCache();
+            UpgradeUtil.clearCacheAndGetNewTable(connection.unwrap(PhoenixConnection.class),
+                    connection.getTenantId()==null?null:connection.getTenantId().getString(),
+                    schema, tableName, systemTransformRecord.getLogicalParentName(), MIN_TABLE_TIMESTAMP);
+            for (TableInfo view : viewsToUpdateCache) {
+                UpgradeUtil.clearCacheAndGetNewTable(connection.unwrap(PhoenixConnection.class),
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/transform/Transform.java:504]: Isn't this potentially a million calls to system catalog? Is there a way we can bulk expire the cache if the views are too numerous? Or just expire without getting the new view PTable until the next time someone needs to query it?

**Review by gjacoby126 — COMMENTED** [review]
Finished first pass

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+        do {
+            if (status.name().equals(record.getTransformStatus())) {
+                return;
+            }
+            Thread.sleep(500); // sleep 1 sec
+            record = Transform.getTransformRecord(record.getSchemaName(), record.getLogicalTableName(), record.getLogicalParentName(), record.getTenantId(), conn);
+            lastStatus = record.getTransformStatus();
+        } while (++nTries < maxTries);
+        try {
+            SingleCellIndexIT.dumpTable("SYSTEM.TASK");
```
[the remark this answers is not in the document]
Comment (gokceni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java:663]: This is very useful in debugging though. This path is reached when the task did not get to the expected state. At that point, it is great to have this log to see what happened

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+    }
+
+    @Test
+    public void testTransformMonitor_immutableTableWithIndex() throws Exception {
+        testTransformTable(true, false, true);
+    }
+
+    @Test
+    public void testTransformMonitor_pausedTransform() throws Exception {
+        testTransformMonitor_checkStates(PTable.TransformStatus.PAUSED, PTable.TaskStatus.COMPLETED);
```
[the remark this answers is not in the document]
Comment (gokceni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java:235]: Yes, task status will be completed. Another task will be created when transform resumes.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/transform/Transform.java (hunk 3 above)
Comment (gokceni) [reply phoenix-core/src/main/java/org/apache/phoenix/schema/transform/Transform.java:504]: Good idea. Let me just expire them without querying.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/tasks/TransformMonitorTask.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+                .setStartTs(startTimestamp)
+                .setEndTs(endTimestamp)
+                .setAccessCheckEnabled(true)
+                .build());
+    }
+
+    @Override
+    public TaskRegionObserver.TaskResult checkCurrentResult(Task.TaskRecord taskRecord)
+            throws Exception {
+        return null;
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/tasks/TransformMonitorTask.java:213]: why is this returning null? Think we need a comment at least please.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+            conn.createStatement().execute("ALTER INDEX " + indexName + " ON " + dataTableName +
+                    " ACTIVE IMMUTABLE_STORAGE_SCHEME=SINGLE_CELL_ARRAY_WITH_OFFSETS, COLUMN_ENCODED_BYTES=2");
+            SystemTransformRecord record = Transform.getTransformRecord(null, indexName, dataTableName, null, conn.unwrap(PhoenixConnection.class));
+            assertNotNull(record);
+            waitForTransformToGetToState(conn.unwrap(PhoenixConnection.class), record, PTable.TransformStatus.COMPLETED);
+            assertMetadata(conn, PTable.ImmutableStorageScheme.SINGLE_CELL_ARRAY_WITH_OFFSETS, PTable.QualifierEncodingScheme.TWO_BYTE_QUALIFIERS, record.getNewPhysicalTableName());
+
+            TransformToolIT.upsertRows(conn, dataTableName, 2, 1);
+
+            conn.createStatement().execute("ALTER INDEX " + indexName + " ON " + dataTableName +
```
[the remark this answers is not in the document]
Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java:441]: I think this comment is still waiting to be implemented.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+            assertEquals(numOfRows+1, countRows(conn, dataTableName));
+
+            // Make sure that we are not accessing the original table. We are supposed to read from the new table above
+            Admin admin = conn.unwrap(PhoenixConnection.class).getQueryServices().getAdmin();
+            TableName hTableName = TableName.valueOf(dataTableName);
+            admin.disableTable(hTableName);
+            admin.deleteTable(hTableName);
+
+            conn.createStatement().execute("ALTER TABLE " + dataTableName +
+                    " SET IMMUTABLE_STORAGE_SCHEME=ONE_CELL_PER_COLUMN, COLUMN_ENCODED_BYTES=0");
```
[the remark this answers is not in the document]
Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java:501]: Ditto

File: phoenix-core/src/main/java/org/apache/phoenix/schema/transform/Transform.java [hunk 9]
```diff
@@ -56,16 +59,45 @@
...
+import static org.apache.phoenix.query.QueryServices.MUTATE_BATCH_SIZE_ATTRIB;
 import static org.apache.phoenix.schema.ColumnMetaDataOps.addColumnMutation;
 import static org.apache.phoenix.schema.MetaDataClient.CREATE_LINK;
+import static org.apache.phoenix.schema.PTable.ImmutableStorageScheme.SINGLE_CELL_ARRAY_WITH_OFFSETS;
 import static org.apache.phoenix.schema.PTableType.INDEX;
 import static org.apache.phoenix.schema.PTableType.VIEW;
 
 public class Transform {
     private static final Logger LOGGER = LoggerFactory.getLogger(Transform.class);
+    private static String transformSelect = "SELECT " +
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/transform/Transform.java:85]: nit: static constant should be TRANSFORM_SELECT

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/tasks/TransformMonitorTask.java (hunk 6 above)
Comment (gokceni) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/tasks/TransformMonitorTask.java:213]: Will add a comment but here is the explanation. IndexRebuildTask uses this function to check if MR job succeeded or not. For this task, we don't need it. The MR job itself updates the task status.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (gokceni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java:441]: @gjacoby126 Please see line 440 (Transform.removeTransformRecord(record, conn.unwrap(PhoenixConnection.class)))
It removes the first one from the table so we are guaranteed that we will get the second one..

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (gokceni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/transform/TransformMonitorIT.java:501]: Line 499 removes the first transform record from the table.

**Review by gjacoby126 — APPROVED** [review]
+1, thanks @gokceni

**Comment by gokceni** [comment]
Merged

**Pull request closed without merging**